### PR TITLE
Autobots, Roll Up! 🤖🚚🎢 - Add basic support for Optimism and non EIP-1559 EVM networks.

### PR DESCRIPTION
**Pull request by 0xDaedalus**
#### As a team we can accomplish anything - Optimus Prime
image

### To Test
- Ethereum & Polygon swaps / sending / dapp interactions should all work as before. (It is probably enough to just test Polygon since it is EIP-1559 compliant just like mainnet).
- Sending assets should be possible via Optimism.
- Swapping on Uniswap should be possible via Optimism.
- The activity panel should reflect transactions made on optimism - and correctly link to etherscan.

### Out of scope
- Gas price preview
- Ledger tx signing
- In-app Swaps
- NFTs

### CODE REVIEW

**Comment by mhluongo** [comment]
This could've been 4 amazing PRs :zany_face: but despite that, I'm here for it :)

**Comment by mhluongo** [comment]
Also apologies in advance, I'm doing this review commit by commit

File: background/services/chain/index.ts [hunk 1]
```diff
@@ -762,6 +763,17 @@ export default class ChainService extends BaseService<Events> {
...
 
+  /**
+   * Estimate the gas needed to make a transaction. Adds 10% as a safety net to
+   * the base estimate returned by the provider.
+   */
+  async estimateGasPrice(network: EVMNetwork): Promise<bigint> {
+    const estimate = await this.providerForNetworkOrThrow(network).getGasPrice()
+
+    // Add 10% more gas as a safety net
+    return BigInt(estimate.toString())
```
Comment (mhluongo) [inline background/services/chain/index.ts:807]: `estimate.toBigInt()`

**Review by mhluongo — COMMENTED** [review]
Loving this! Left a bunch of code-level comments and giving it a test drive now

**Comment by mhluongo** [comment]
After sending a token (ETH) to myself on Polygon, then switching to Optimism, the Redux store blows up

Screenshot from 2022-08-11 18-10-32

**Comment by mhluongo** [comment]
Assuming `transactionConstruction.signedTransaction.gasPrice` should be a `bigint`

**Comment by mhluongo** [comment]
@0xDaedalus I ended up debugging this and I believe I fixed the bug.

I do think that `as unknown as SignedTransaction` that hid the issue needs to be removed before this is merged.

**Comment by 0xDaedalus** [comment]
Thank you for the amazing feedback and debugging!  Will be diving into this shortly ❤️

File: background/services/enrichment/types.ts [hunk 2]
```diff
@@ -100,6 +113,24 @@ export type EnrichedEIP1559TransactionRequest = EIP1559TransactionRequest & {
...
+}
+
+export type EnrichedTransactionRequest =
+  | EnrichedEIP1559TransactionRequest
+  | EnrichedLegacyTransactionRequest
+
+export type EIP1559TransactionRequestWithFrom =
+  | Partial<EIP1559TransactionRequest> & { from: string }
+
+export type LegacyEVMTransactionRequestWithFrom =
```
[the remark this answers is not in the document]
Comment (0xDaedalus) [reply background/services/enrichment/types.ts]: Good question - I think we always do - and this type is existed (and was factored out into its own type as part of this PR) to tell the type checker that the object it is typing can be a `Partial` but _must_ have the `from` property.  I'll go ahead and rename the type to make that a bit more clear.

File: background/networks.ts [hunk 3]
```diff
@@ -318,3 +343,27 @@ export function toHexChainID(chainID: string | number): string {
   }
   return `0x${BigInt(chainID).toString(16)}`
 }
+
+export function isEIP1559TransactionRequest(
+  transactionRequest: AnyEVMTransaction | Partial<TransactionRequestWithFrom>
+): transactionRequest is EIP1559TransactionRequest {
+  if (
```
[the remark this answers is not in the document]
Comment (0xDaedalus) [reply background/networks.ts:360]: Excellent point

File: background/main.ts [hunk 4]
```diff
@@ -31,7 +31,11 @@ import {
 } from "./services"
 
 import { HexString, KeyringTypes } from "./types"
-import { AnyEVMTransaction, SignedEVMTransaction } from "./networks"
+import {
+  AnyEVMTransaction,
+  SignedEVMTransaction,
+  TransactionRequest,
+} from "./networks"
```
[the remark this answers is not in the document]
Comment (0xDaedalus) [reply background/main.ts]: ✅  https://github.com/tallycash/extension/issues/1925

File: background/main.ts [hunk 5]
```diff
@@ -145,6 +153,7 @@ const devToolsSanitizer = (input: unknown) => {
 const reduxCache: Middleware = (store) => (next) => (action) => {
   const result = next(action)
   const state = store.getState()
+  ;(window as any).store = store
```
[the remark this answers is not in the document]
Comment (0xDaedalus) [reply background/main.ts]: ✅

File: background/services/chain/index.ts [hunk 6]
```diff
@@ -343,6 +346,76 @@ export default class ChainService extends BaseService<Events> {
...
+        ) {
+          gasEstimationError = anyError.error ?? "unknown transaction error"
+        }
+      }
+    }
+
+    // We use the estimate as the actual limit only if user did not specify the
+    // gas explicitly or if it was set below the minimum network-allowed value.
+    if (
+      typeof estimatedGasLimit !== "undefined" &&
```
[the remark this answers is not in the document]
Comment (0xDaedalus) [reply background/services/chain/index.ts:412]: I believe that this comment is in reference to an intermediate commit - the 6m fallback was a fallback before `estimateGas` was a thing and is no longer present.

File: background/services/chain/index.ts (hunk 1 above)
Comment (0xDaedalus) [reply background/services/chain/index.ts]: Resolved by sir Luongo himself

File: background/services/chain/index.ts [hunk 7]
```diff
@@ -343,6 +346,76 @@ export default class ChainService extends BaseService<Events> {
...
+   */
+  async populatePartialLegacyEVMTransactionRequest(
+    network: EVMNetwork,
+    partialRequest: EnrichedLegacyTransactionSignatureRequest
+  ): Promise<{
+    transactionRequest: EnrichedLegacyTransactionRequest
+    gasEstimationError: string | undefined
+  }> {
+    // Basic transaction construction based on the provided options, with extra data from the chain service
+    const transactionRequest: EnrichedLegacyTransactionRequest = {
```
[the remark this answers is not in the document]
Comment (0xDaedalus) [reply background/services/chain/index.ts:368]: Definitely - this was a copy/paste & modify over from the equivalent EIP1559 function so definitely good to fix the cruft while we're in here

File: background/services/chain/utils.ts [hunk 8]
```diff
@@ -120,30 +155,75 @@ export function eip1559TransactionRequestFromEthersTransactionRequest(
   }
 }
 
+function legacyEVMTransactionRequestFromEthersTransactionRequest(
+  transaction: EthersTransactionRequest
+): Partial<LegacyEVMTransactionRequest> {
+  // TODO What to do if transaction is not EIP1559?
```
[the remark this answers is not in the document]
Comment (0xDaedalus) [reply background/services/chain/utils.ts:160]: Indeed - removed it from the original function as well

File: background/services/chain/utils.ts [hunk 9]
```diff
@@ -120,30 +155,75 @@ export function eip1559TransactionRequestFromEthersTransactionRequest(
...
+        ? parseInt(transaction.nonce.toString(), 16)
+        : undefined,
+    value:
+      typeof transaction.value !== "undefined"
+        ? BigInt(transaction.value.toString())
+        : undefined,
+    chainID: transaction.chainId?.toString(16),
+    gasLimit:
+      typeof transaction.gasLimit !== "undefined"
+        ? BigInt(transaction.gasLimit.toString())
```
[the remark this answers is not in the document]
Comment (0xDaedalus) [reply background/services/chain/utils.ts:178]: Compiler doesn't like that unfortunately

File: background/services/chain/index.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (mhluongo) [reply background/services/chain/index.ts:413]: Yep, resolved!

**Review by mhluongo — APPROVED** [review]
Looking good :sunglasses:

File: background/services/chain/utils.ts [hunk 10]
```diff
@@ -120,30 +156,74 @@ export function eip1559TransactionRequestFromEthersTransactionRequest(
...
 ): EthersTransaction {
-  const baseTx = {
+  const baseTx: EthersTransaction = {
     nonce: Number(tx.nonce),
-    maxFeePerGas: tx.maxFeePerGas ? BigNumber.from(tx.maxFeePerGas) : undefined,
-    maxPriorityFeePerGas: tx.maxPriorityFeePerGas
-      ? BigNumber.from(tx.maxPriorityFeePerGas)
-      : undefined,
     to: tx.to,
-    from: tx.from,
```
Comment (Shadowfiend) [inline background/services/chain/utils.ts:133]: Why only return `from` for EIP1559?

Comment (0xDaedalus) [reply background/services/chain/utils.ts:133]: According to ethersproject - [`from` is not an allowed key for for legacy transaction serialization](https://github.com/ethers-io/ethers.js/blob/master/packages/transactions/src.ts/index.ts#L105)

**Pull request merged**